Lib/shutil.py

```diff
@@ -813,6 +813,10 @@ def move(src, dst, copy_function=copy2):
             if _destinsrc(src, dst):
                 raise Error("Cannot move a directory '%s' into itself"
                             " '%s'." % (src, dst))
+            if not os.access(src, os.W_OK) and os.listdir(src):
```

Comment:
Renaming a file in POSIX requires write access to the file's current parent directory and to the destination directory. When *renaming a directory*, the operation additionally requires write access to the renamed directory if, and only if, the current parent directory and the destination directory are different. This is because the ".." entry in the renamed directory has to be modified to reference the new parent directory. Write access to the renamed directory is not required if the current parent directory and the destination directory are the same, because the ".." entry does not have to be modified in this case. But just unlinking an empty directory, as opposed to renaming it, only requires write access to the parent directory. 

An exception is that a directory with the [i]mmutable attribute cannot be renamed or unlinked. There is no specific test for this in POSIX, except that the error code will probably be `EPERM` instead of `EACCES`. Linux has a `statx()` system call to check for the `STATX_ATTR_IMMUTABLE` attribute. Testing for write access on the target directory itself works indirectly because of course write access is also denied for an immutable directory.

---

In Windows, renaming is implemented via `MoveFileExW()`, which requires synchronize, read-attributes, and delete access to the renamed file or directory. The last two can be provided directly by the security of the renamed file, or indirectly if its parent directory grants list-directory (i.e. read-data) and delete-child access. Synchronize access has to be granted directly by the renamed file. For the destination directory, a rename requires synchronize access plus either add-file (i.e. write-data) or add-subdirectory (i.e. append-data) access, respectively if a file or directory is being renamed into the destination. 

The analog to the Unix immutable attribute is the readonly attribute, except readonly only prevents modifying *data* streams in a file or directory (a directory can have named data streams) and deleting a file or empty directory. It doesn't prevent modifying metadata. It doesn't prevent getting delete access itself, so renaming is still possible. Also, it doesn't pertain to the *index* stream in a directory, i.e. it does not prevent obtaining add-file or add-subdirectory access to the destination directory.

Reply:
Thanks for the insight. However, this function (`shutil::move`) doesn't rename on the file system level; it leverages shutil's `copytree` and `rmtree` functions instead.

This PR simply addresses the inability for `shutil::rmtree` to remove the source directory during the renaming process. This is a concern for both POSIX and Windows: `rmtree` cannot remove the source directory if the user lacks access to the source directory.

Reply:
Yes, it's clear that checking for write access on `src` is about unlinking the *immediate* contents of `src`, but I wanted to clarify the permission-based reasons that can make the `os.rename()` call fail in `shutil.move()`. What troubled me was your claim that being able to unlink an empty directory without write access to it is a "weird OS quirk". That seems like a misunderstanding of POSIX filesystem permissions, for which the right to unlink a file or directory is controlled by the parent directory, unless the file is flagged [i]mmutable (e.g. via `chattr` in Linux). IMO, the quirk, which possibly leads to misunderstanding, is that write access to a directory is required in order to rename it to a new parent directory, which is explained by the need to modify its ".." entry.

Note that if `src` is a directory, write access to it doesn't imply `rmtree(src)` will succeed, since subdirectories in the tree may lack write access. It also has nothing to do with the success of `os.rmdir(src)` when it's empty, as the last operation of `rmtree(src)`. The latter depends on write access to the parent of `src`.

Personally, I'd prefer for the fallback copy/delete behavior to be restricted to just `EXDEV` failures due to a cross-filesystem rename. If `os.rename()` (or rather, `os.replace()`) fails for another reason, I'd prefer to just let `move()` fail. Hopefully that catches most of the permission errors, such as the quirk in POSIX with needing write access to a directory in order to rename it to a new parent directory. Of course, for the `EXDEV` case, it depends on the order in which the system checks permissions compared to the cross-filesystem status. If the latter is checked first, then permissions may still be a problem. But what are we going to do for LBYL in the `EXDEV` case, recursively check all subdirectories in the entire `src` tree for write access? That's pushing TOCTOU to an extreme.

---

For Windows, checking `os.access()` is currently pointless for a directory. This function does not actually implement an access check. It just checks the readonly file attribute of a non-directory file. Directories are excluded because the readonly attribute only prevents writing to a data stream, not an index stream, so it has nothing to do with adding and removing files and subdirectories. Also, the filesystem security model is very different in Windows. A directory can grant delete-child access to allow delete/rename access to its immediately contained files. But a file can also be individually granted delete/rename access. If we implemented `os.access()` for real, we could combine generic-write and delete-child access as POSIX-like `W_OK` access. But there would still be a fundamental difference that a file can be granted delete/rename access individually. Also, write access to a directory index is split into separate add-file and add-subdirectory rights. A directory might allow one and not the other, but we'd have to test for both with POSIX `W_OK` access. Thus lack of POSIX-like write access to a directory would not imply anything about whether deleting its files or subdirectories should succeed.

Reply:
To repeat more succinctly, checking for write access on `src` is inadequate in POSIX because the caller may lack write access to subdirectories and the parent directory, and it's completely meaningless in Windows.

Reply:
Appreciate the clarification. To address your concerns:

1. Thanks for clarifying the empty directory "quirk" with regards to parent permissions, and it's great that you brought up immutability as an exception. I've added an additional check for immutability in this PR, since it doesn't seem like `shutil` had previously acknowledged this possibility.
2. Going back, when I used the word "quirk", I meant it more as a shutil edge case than an actual OS concern. We need the `os.listdir(src)==False` check just because it's the only situation where a non-writable child directory can be deleted...which, you're right, of course is related to the parent's file data. Again, this is only a quirk because `rmtree` is being used as a renaming crutch.
3. `os.access(src, os.W_OK)==False` means that rmtree *will* fail (with the exception of (2)), and you're correct that that rmtree could also fail when subdirectories lack write access. However, the latter isn't a concern in this part of the code. The converse situation (`os.access(src, os.W_OK)==True` but with unknown `src` subdirectories) is already avoided since `os.rename()` would succeed above. The same logic applies for Windows. Even though the access check is different, the inability to remove *any* file implies an `rmtree` failure.

Overall, I think we both agree that the fallback copy/delete behavior can be improved. The fact that all this cross-system file logic is being handled within an `except OSError` is concerning to me too.

But 2-line PR is meant more as a bugfix :) I'd be happy to discuss some refactoring with you, but that's probably best left for another PR.

Reply:
Hello eryksun. 

https://stackoverflow.com/questions/4495767/create-a-directory-and-get-the-handle-by-issuing-one-irp
A question about this.
Can you show me an example of how to do so? I tried this myself with win32 APIs are it fails.